TabularMSA: __getitem__, loc, iloc

ebolyen opened the pull request:
Still need to add docs

@jairideout would you mind giving this a first pass?

fixes #1102 
fixes #1086 
fixes #1088

ebolyen commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,203 @@
+
+        if type(indexable) is tuple:
+            if len(indexable) > 2:
+                raise ValueError("TODO")
```
Crap... I'll fix this.

ebolyen commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,203 @@
+            return indexable
+
+        if indexable.dtype == bool and len(indexable) != self._obj.shape[axis]:
+            raise IndexError()
```
Same problem here

ebolyen commented:
and sprinkle in some flake8 fixes...

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import pandas as pd
```
It begins.

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+
+        if type(indexable) is tuple:
+            if len(indexable) > 2:
+                raise ValueError("Can only slice on two axis. Tuple is length:"
```
*axis* -> *axes*

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+        seq_index = self._convert_ellipsis(seq_index)
+        pos_index = self._convert_ellipsis(pos_index)
+
+        if not hasattr(seq_index, '__iter__') and (seq_index is Ellipsis or
```
Is the `Ellipsis` check necessary since you're calling `_convert_ellipsis(seq_index)`?

ebolyen replied:
nope! Thanks!

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+        if isinstance(indexable, np.ndarray):
+            pass
+        elif hasattr(indexable, '__iter__'):
+            indexable = np.asarray(list(indexable))
```
Is `list` necessary here?

ebolyen replied:
Yeah, generators are possible, which while meaningless to `loc` are meaningful to `iloc`. So to share logic, we need to convert.

jairideout replied:
kk

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+
+
+class TabularMSAILoc(_Indexing):
+    def is_scalar(self, indexable, axis):
```
Can you make `_Indexing` an ABC with these abstract methods? That'll help make the code self-documenting (it wasn't obvious to me why these methods were being defined here).

ebolyen replied:
Yep

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+            if not self._has_fancy_index():
+                # prevents unfriendly errors
+                raise TypeError("Cannot provide a tuple to the first axis of"
+                                " `loc` unless the msa's `index` is a"
```
*msa* -> *MSA*

jairideout commented on skbio/alignment/_indexing.py:
```diff
@@ -0,0 +1,206 @@
+                                " `pd.MultiIndex`.")
+            elif self.is_scalar(indexable[0], axis=0):
+                # prevents unreasonable results
+                # pd.Series.loc[('a', 0), ('b', 1)] would be interepreted as
```
*interepreted* -> *interpreted*

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -317,6 +339,43 @@ def __init__(self, sequences, metadata=None, positional_metadata=None,
+        """
+
+        if sequences is NotImplemented:
+            sequences = self._seqs
```
Should sequences get shallow-copied in the constructor? `metadata` and `positional_metadata` are being shallow-copied in the constructor and `index` is immutable but sequences aren't being copied.

ebolyen replied:
:+1: I think you are right.

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -639,6 +698,64 @@ def __deepcopy__(self, memo):
+        return self.iloc[indexable]
+
+    # Helpers for TabularMSAILoc and TabularMSALoc
+    def _get_sequence_(self, i):
```
Can you rename this `_get_sequence_iloc_` to mirror `_get_sequence_loc_`?

ebolyen replied:
:+1:

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -639,6 +698,64 @@ def __deepcopy__(self, memo):
+    def _get_sequence_(self, i):
+        return self._seqs.iloc[i]
+
+    def _slice_sequences_(self, i):
```
Same comment: `_slice_sequences_iloc_`

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -1509,8 +1626,7 @@ def _get_join_index(self, other, how):
 
     def _get_sequence_for_join(self, label):
         if label in self.index:
-            # TODO: use .loc when it is implemented.
```
Can you grep for other TODOs under `skbio.alignment` and `skbio.io`? I think I marked a couple places where `loc`/`iloc`/`__getitem__` should be used.

ebolyen replied:
will do

ebolyen replied:
Strangely, after looking through all of the TODO/todo in our package, none of them came up mentioning loc/iloc/getitem

jairideout replied:
Thanks for checking, you must have fixed the relevant TODOs already

jairideout commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -901,6 +899,876 @@ def test_with_index(self):
+                                         positional_metadata={0: [1, 2, 3]},
+                                         index=[True, False]))
+
+    def test_bool_index_second_second(self):
```
`test_bool_index_second_second` -> `test_bool_index_second_axis`

ebolyen replied:
Thanks!

jairideout commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -901,6 +899,876 @@ def test_with_index(self):
+           This does mean the iloc via a fancy index and simple scalar must
+           work correctly.
+        """
+        # This will be overriden for TestLoc because the first axis are labels
```
put this in `TestILoc`?

ebolyen replied:
I ideally want to share those between the second axis of `TestLoc` as well.

jairideout replied:
makes sense

jairideout commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -901,6 +899,876 @@ def test_with_index(self):
+            self.get(msa, (..., 3))
+
+
+class TestGetItem(SharedIndexTests, unittest.TestCase):
```
What about having this subclass `TestILoc`? Then `test_uses_iloc_not_loc` can be removed and `__getitem__` will get tested the same as `iloc`.

ebolyen replied:
I had it like that originally, but some of the tests use `axis` which can't be expressed with `getitem`. I can probably pull those out into a fourth class for `iloc` specific tests.

jairideout replied:
If it's quick/easy to do, probably worth it. If not, okay to leave as-is

jairideout commented on skbio/sequence/tests/test_sequence.py:
```diff
@@ -85,6 +85,11 @@ def test_concat_on_subclass(self):
         self.assertEqual(result, SequenceSubclass("123123"))
 
+    def test_concat_on_empty_iterator(self):
+        result = SequenceSubclass.concat((_ for _ in []))
```
`numpy.concatenate` raises a `ValueError` in this case. Should we match that behavior?

ebolyen replied:
numpy wouldn't always have a dtype. I think we can avoid that because the output is always predictable being a classmethod instead of a function. 

I also added this behaviour because it makes the implementation simpler as I don't need to special case an empty slice situation when pulling out a column.

If we feel strongly about raising an error that is fine, but I think that 
`Sequence.concat([]) -> Sequence("")` is a logically consistent result.

jairideout replied:
Makes sense, thanks for the details

jairideout commented:
@ebolyen this looks great! I really like the factoring, the code is very self-documenting and maintainable. Just had some mostly minor comments. Can you verify coverage locally? (I don't trust coveralls)

ebolyen commented:
Coverage is at :100:*%* locally. I've got a command that watches the test file and reruns coverage on write.

ebolyen commented:
@jairideout addressed your comments, not sure what is taking travis so long, but I should have fixed the Py2 errors as well.

Docs will be in the next commit.

gregcaporaso commented:
It's release day, doesn't Travis usually take these days off?

On Wed, Dec 9, 2015 at 10:50 AM, Evan Bolyen <notifications@github.com>
wrote:

> @jairideout <https://github.com/jairideout> addressed your comments, not
> sure what is taking travis so long, but I should have fixed the Py2 errors
> as well.
>
> Docs will be in the next commit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1224#issuecomment-163337884>.
>

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -318,6 +757,44 @@ def __init__(self, sequences, metadata=None, positional_metadata=None,
+        if sequences is NotImplemented:
+            sequences = self._seqs
+
+        sequences = [copy.copy(s) for s in sequences]
```
This should go in `__init__` so that copy behavior is consistent whether you're creating a new `TabularMSA` object or a `TabularMSA` derived from another MSA. Can you add a unit test for `__init__` performing a shallow copy of sequences and mention this behavior in the docs?

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        ACGT
+        A-GT
+        AC-T
+        ACGA
```
Can you show the index below this line?

```python
>>> msa.index
```

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        alignment has no biological meaning and many operations defined for the
+        MSA's sequence `dtype` would be meaningless.
+
+        When we slice both axis by a scalar, operations are applied left to
```
*axis* -> *axes*

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        ACGT
+        A-GT
+
+        Remember that `iloc` can always be used to differentiate sequences.
```
*...to differentiate sequences with duplicate labels.*

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        alignment has no biological meaning and many operations defined for the
+        MSA's sequence `dtype` would be meaningless.
+
+        When we slice both axis by a scalar, operations are applied left to
```
*axis* -> *axes*

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -640,6 +1117,77 @@ def __deepcopy__(self, memo):
+
+        To slice by labels, use ``loc``.
+
+        See Also
```
put See Also before Notes

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        If the slice operation results in a ``TabularMSA`` without any
+        sequences, the MSA's ``positional_metadata`` will be unset.
+
+        See Also
```
put See Also before Notes

jairideout commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -245,12 +246,450 @@ def index(self):
+        `seq_idx` to indicate the slicing operations to perform on each
+        component index.
+
+        See Also
```
put See Also before Notes

jairideout commented:
@ebolyen a few minor doc-related comments and shallow copy sequences in `__init__`. Ready for merge after that!
